### PR TITLE
[3751] Flatten page params for public/v1

### PR DESCRIPTION
**Pull request by asmega**
### Context
https://trello.com/c/d6pSLXpC/53-m-flatten-page-parameters

Swapping out Kaminari for Pagy, below is the memory usage for the relevant gems being swapped in and out.

#### Before
```
allocated memory by gem
-----------------------------------
   4013688  activerecord-6.0.3.2
   2817120  awesome_print-1.8.0
   2016402  activesupport-6.0.3.2
   1068337  activemodel-6.0.3.2
    652616  semantic_logger-4.6.1
    449136  jsonapi-serializable-0.3.1
    376982  json-2.3.0
    257273  rack-mini-profiler-2.0.4
    187320  rails_semantic_logger-4.4.4
    150055  lib
    149584  teacher-training-api/app
     42443  actionpack-6.0.3.2
     30880  dry-container-0.7.2
     28784  jsonapi-renderer-0.2.2
     18816  aasm-5.1.1
     17114  timecop-0.9.1
     13378  rack-2.2.3
     11264  discard-1.2.0
      6624  tzinfo-1.2.7
      4352  concurrent-ruby-1.1.7
      2767  api-pagination-4.8.2 <--------------
      1574  jsonapi-rails-0.4.0
      1398  sentry-raven-3.0.2
      1360  audited-4.9.0
      1192  actionview-6.0.3.2
       856  other
       824  skylight-core-4.3.1
       680  kaminari-core-1.2.1 <--------------
       309  teacher-training-api/lib
       272  request_store-1.5.0
       200  kaminari-activerecord-1.2.1 <--------------
       112  railties-6.0.3.2
        80  puma-4.3.5
```
#### After
```
allocated memory by gem
-----------------------------------
   4001136  activerecord-6.0.3.2
   2911176  awesome_print-1.8.0
   1963374  activesupport-6.0.3.2
   1068337  activemodel-6.0.3.2
    670935  semantic_logger-4.6.1
    449136  jsonapi-serializable-0.3.1
    376223  json-2.3.0
    257505  rack-mini-profiler-2.0.4
    187320  rails_semantic_logger-4.4.4
    150048  teacher-training-api/app
    148183  lib
     30880  dry-container-0.7.2
     29248  jsonapi-renderer-0.2.2
     27958  actionpack-6.0.3.2
     18816  aasm-5.1.1
     17114  timecop-0.9.1
     13218  rack-2.2.3
     11264  discard-1.2.0
      6624  tzinfo-1.2.7
      5048  concurrent-ruby-1.1.7
      1848  actionview-6.0.3.2
      1592  audited-4.9.0
      1574  jsonapi-rails-0.4.0
      1398  sentry-raven-3.0.2
      1240  pagy-3.8.3 <--------------
       824  skylight-core-4.3.1
       504  request_store-1.5.0
       309  teacher-training-api/lib
       232  other
       112  railties-6.0.3.2
        40  i18n-1.8.3
        40  puma-4.3.5
```

### Changes proposed in this pull request

- Swap out Kaminari for more modern Pagy gem
- public/v1 endpoints now paginate with `page` and `per_page` params.
These are no longer nested withing `page`
- All previous endpoints unchanged and still continue to use
`page[page]` and `page[per_page]`
- Remove headers related to pagination in responses as these do not
appear to be used anywhere
- Removed pagination from private/v1 subjects enpoints as it never hit
the page limit therefore pagination is not needed
- Reimplemented `paginate` to delegate to Pagy for backwards
compatability

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by asmega** [comment]
[[M] Flatten page parameters](https://trello.com/c/d6pSLXpC/3751-m-flatten-page-parameters)

File: app/controllers/api/v1/providers_controller.rb [hunk 1]
```diff
@@ -31,6 +30,10 @@ def index
 
     private
 
+      def per_page
+        params[:per_page] || 100
```
Comment (defong) [inline app/controllers/api/v1/providers_controller.rb:34]:
```suggestion
        (params[:per_page] || 100).to_i
```

https://github.com/DFE-Digital/teacher-training-api/blob/7226c70ec1e1aa34cd0a4294a18b30c86af77cc6/app/controllers/api/v1/courses_controller.rb#L41

File: app/controllers/api/v1/courses_controller.rb [hunk 2]
```diff
@@ -38,6 +37,10 @@ def index
 
     private
 
+      def per_page
```
Comment (darokel) [inline app/controllers/api/v1/courses_controller.rb:40]: How comes we're overriding the `per_page` method here with essentially the same logic as the one defined in https://github.com/DFE-Digital/teacher-training-api/pull/1547/files#diff-100c2681a1f561f361ad2a624c23c8feR22?

File: app/controllers/api/v2/application_controller.rb [hunk 3]
```diff
@@ -23,6 +25,31 @@ def jsonapi_404
...
+      def paginate(scope, per_page:)
+        _pagy, paginated_records = pagy(scope, items: per_page, page: page)
+
+        paginated_records
+      end
+
+      def per_page
+        params[:page] ||= {}
+
+        [(params[:page][:per_page] || default_per_page).to_i, max_per_page].min
```
Comment (defong) [inline app/controllers/api/v2/application_controller.rb:37]:
```suggestion
        [(params.dig(:page, :per_page) || default_per_page).to_i, max_per_page].min
```

File: app/controllers/api/v2/application_controller.rb [hunk 4]
```diff
@@ -23,6 +25,31 @@ def jsonapi_404
...
+        100
+      end
+
+      def max_per_page
+        100
+      end
+
+      def page
+        params[:page] ||= {}
+        params[:page][:page] || 1
```
Comment (defong) [inline app/controllers/api/v2/application_controller.rb:50]:
```suggestion
        (params.dig(:page, :page) || 1).to_i
```

File: app/controllers/api/public/v1/application_controller.rb [hunk 5]
```diff
@@ -2,11 +2,37 @@ module API
...
+
         rescue_from ActiveRecord::RecordNotFound, with: :jsonapi_404
 
         def jsonapi_404
           render jsonapi: nil, status: :not_found
         end
+
+      private
+
+        def paginate(scope)
```
Comment (darokel) [inline app/controllers/api/public/v1/application_controller.rb:15]: I noticed that these methods are being duplicated into all `application_controller.rb` files for the old api versions and the new public one. Is it worth just putting these methods into a `Paginatable` module and including that into those application controller classes?

File: app/controllers/api/v3/application_controller.rb [hunk 6]
```diff
@@ -1,31 +1,41 @@
...
+
+        paginated_records
+      end
+
+      def per_page
+        params[:page] ||= {}
 
-        if params[:page][:per_page].to_i > Kaminari.config.default_per_page
-          return if allowed_to_disable_pagination?
+        [(params[:page][:per_page] || default_per_page).to_i, max_per_page].min
```
Comment (defong) [inline app/controllers/api/v3/application_controller.rb:25]:
```suggestion
        [(params.dig(:page, :per_page) || default_per_page).to_i, max_per_page].min
```

File: app/controllers/api/v3/application_controller.rb [hunk 7]
```diff
@@ -1,31 +1,41 @@
...
+        100
       end
 
-      # Override if you want to allow an endpoint to disable pagaination
-      # e.g. by checking params
-      def allowed_to_disable_pagination?
-        false
+      def page
+        params[:page] ||= {}
+        params[:page][:page] || 1
```
Comment (defong) [inline app/controllers/api/v3/application_controller.rb:38]:
```suggestion
        (params.dig(:page, :page) || 1).to_i
```

File: spec/requests/api/v3/courses/index_spec.rb [hunk 8]
```diff
@@ -208,10 +191,9 @@
           json_response = JSON.parse(response.body)
           course_ids << json_response.dig("data", 0, "id")
 
-          get page_3
-
-          json_response = JSON.parse(response.body)
-          course_ids << json_response.dig("data", 0, "id")
+          expect {
+            get page_3
+          }.to raise_error(Pagy::OverflowError)
```
Comment (defong) [inline spec/requests/api/v3/courses/index_spec.rb:196]: So when something is out of bounds it raises error...

Maybe add it here as well?
https://github.com/DFE-Digital/teacher-training-api/blob/66fc56992741bb605b9dd932505002cc1ec2dc0c/config/application.rb#L38-L42

File: spec/requests/api/v3/providers/courses/index_spec.rb [hunk 9]
```diff
@@ -123,9 +123,6 @@ def perform_request
             "meta" => {
               "count" => 1,
             },
-            "links"  => {
-              "last" => "/api/v3/recruitment_cycles/2020/providers/#{provider.provider_code}/courses?page%5Bpage%5D=1",
```
Comment (defong) [inline spec/requests/api/v3/providers/courses/index_spec.rb:127]: How come this changed?

**Review by defong — COMMENTED** [review]
Just a couple of questions...

File: app/controllers/api/v1/providers_controller.rb (hunk 1 above)
Comment (asmega) [reply app/controllers/api/v1/providers_controller.rb:34]: with your change `nil` would get converted to `0` the original would return `100`

Comment (defong) [reply app/controllers/api/v1/providers_controller.rb:34]: So is the others right then?

File: spec/requests/api/v3/providers/courses/index_spec.rb (hunk 9 above)
Comment (asmega) [reply spec/requests/api/v3/providers/courses/index_spec.rb:127]: seems to be a feature of `api-pagination` we has now been removed. If we're not using it i say we should remove it

File: app/controllers/api/v1/courses_controller.rb (hunk 2 above)
Comment (asmega) [reply app/controllers/api/v1/courses_controller.rb:40]: v1 seems to work differently compared to the other endpoints. If you actually look at all the v1 endpoints they never actually call `paginate`. `per_page` is used as the activerecord/sql `limit`

File: app/controllers/api/public/v1/application_controller.rb (hunk 5 above)
Comment (asmega) [reply app/controllers/api/public/v1/application_controller.rb:15]: I'd rather opt for this duplication. I don't see the benefit of having this as shared code. The pagination code is fairly slim. For public/v1 we could probably drop the `paginate` method soon as i only added it for backwards compatibility as they is how the current code calls it

File: app/controllers/api/v1/providers_controller.rb (hunk 1 above)
Comment (asmega) [reply app/controllers/api/v1/providers_controller.rb]: I see, i've tweaked this and a few of the others

File: spec/requests/api/v3/courses/index_spec.rb (hunk 8 above)
Comment (asmega) [reply spec/requests/api/v3/courses/index_spec.rb:196]: done

File: app/controllers/api/v3/application_controller.rb (hunk 6 above)
Comment (asmega) [reply app/controllers/api/v3/application_controller.rb]: done

File: app/controllers/api/v3/application_controller.rb (hunk 7 above)
Comment (asmega) [reply app/controllers/api/v3/application_controller.rb]: done

File: app/controllers/api/v2/application_controller.rb (hunk 4 above)
Comment (asmega) [reply app/controllers/api/v2/application_controller.rb]: done

File: app/controllers/api/v2/application_controller.rb (hunk 3 above)
Comment (asmega) [reply app/controllers/api/v2/application_controller.rb]: done

File: app/controllers/api/public/v1/application_controller.rb (hunk 5 above)
Comment (darokel) [reply app/controllers/api/public/v1/application_controller.rb:15]: yeah that's fair

**Review by darokel — APPROVED** [review]

**Review by defong — APPROVED** [review]

**Pull request merged**